### PR TITLE
Refactor Map Caching Changes

### PR DESCRIPTION
**Pull request by jlewis91**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?

- Removes duplicated computation in dict comprehension
- Enables `any` short-circuiting
- Eliminates nested for loop

## Why is this PR important?

- Small tweaks to the codebase
- First contribution experience for a new contributor ;)

### CODE REVIEW

File: src/prefect/core/flow.py [hunk 1]
```diff
@@ -883,21 +883,26 @@ def _run_on_schedule(
...
-                        new = prefect.engine.state.Mapped(map_states=[])
-                        for sub_state in s.map_states:
-                            new.map_states.append(
-                                sub_state if sub_state.is_cached() else None
-                            )
-                        cached_tasks.update({t: new})
+                        cached_sub_states = [
+                            sub_state
+                            for sub_state in s.map_states
+                            if sub_state.is_cached()
```
Comment (cicdw) [inline src/prefect/core/flow.py]: this should have an `else None` condition for those mapped states which are _not_ cached, to ensure that the order / length of the mapped states is consistent.

Comment (jlewis91) [reply src/prefect/core/flow.py]: Nice catch. Fixed. One question...why is this not caught by the test? Would it make sense to add something around this?

Comment (cicdw) [reply src/prefect/core/flow.py]: Yea, this change only affects the situation where some of the "mapped children" are cached and some aren't - my test has _both_ cached.  I will add a test for this!

**Comment by cicdw** [comment]
@jlewis91 if you'd like to add your name to the Contributors list in the CHANGELOG, feel free to!  I'll leave this open for a bit.

**Pull request merged**